pyk/src/pyk/kast/manip.py

```diff
@@ -749,7 +760,8 @@ def build_rule(
       - `rule`: A `KRule` with variable naming conventions applied so that it should be parseable by K frontend.
       - `var_map`: The variable renamings that happened to make the claim parseable by K frontend (which can be undone to recover original variables).
     """
-    init_constraints = list(init_constraints)
+    init_constraints = [bool_to_ml_pred(simplify_bool(ml_pred_to_bool(c))) for c in init_constraints]
+    final_constraints = [bool_to_ml_pred(simplify_bool(ml_pred_to_bool(c))) for c in final_constraints]
     final_constraints = [c for c in final_constraints if c not in init_constraints]
```

Comment:
There are subtle differences between the initial and the final constraints that mean that the original filtering does not work as intended. An example of such a subtle differences is having
```
#Not ( CALLER_ID #Equals 645326474426547203313410069153905908525362434349 }
```
in the initial constraints versus
```
{ true #Equals CALLER_ID =/=Int 645326474426547203313410069153905908525362434349 }
```
in the final constraints.

By moving to `Bool`, simplifying there, and then moving back to `ML`, we iron out these subtleties.

Reply:
This is a pretty hacky way to achieve that. I would prefer if instead we just normalized them.

At the very least, you can extract a new function called `normalize_ml_pred(KInner) -> KInner`, which calls the sequence of functions you're going for, and at least then the hackiness can be unit-tested and isolated so we can make an improved routine for it later.

Better would be to just implement `simplify_ml_pred(...)` which directly normalizes these things how we want, rather than relying on translating to bool.

Reply:
I guess the way to start is:

```
def normalize_ml_pred(pred: KInner) -> KInner:
    ...
```

Then you can add a unit-testing harness and put in some inputs that exercise the various cases that `ml_pred_to_bool` handles. You can probably just use all teh inputs that the tests for `ml_pred_to_bool` uses.

Reply:
Yes, I'm very happy to do that, I was thinking of that, but then I thought that that would result in a duplication with respect to `simplify_bool`, and also in the context of that we want to move away from ML connectives.

Reply:
I would like to go with `simplify_ml_pred`, that was my second option, just taking the appropriate cases from `simplify_bool`.